### PR TITLE
Updating price, on_hand, on_demand validations for product import

### LINKED ISSUE #9978: [Import] Some invalid input is not caught by import validation
## Description

This is a follow up of #9911.

During import of products and variants some invalid input is not caught by validation. This should be improved to increase usability of the feature.

After import:
Table:
image
Messages:
image

Regarding :x: 
- Line 4: Shifted to the left but imported correctly :arrow_right: #5961
- Line 13: Unit is invalid and the variant is not created even though it is listed as 'will be created' during import --> this should be listed in the error list and import should be stopped
- Line 18: Price is invalid and so the price is set to 0 --> there could be an error if price is invalid and import should be stopped
- Line 20: On hand is invalid and would be set to 0 if on demand = 0 --> there could be an error if on hand is invalid and import should be stopped
- Line 23: On demand is invalid (not empty or 0) and so is set to on demand = 1 --> there could be an error if on demand is invalid and import should be stopped

Regarding :question: 
- Line 3: Text could be optimized to check for misspellings too. This can be done in Transifex.
- Line 6: Same text as for tax categories and shipping categories should be used here.
- Line 10: see line 6.
- Line 16: I don't understand the error message. Maybe it means: Variant_unit_name must be identical for all variants of a product. 

## Expected Behavior

See above.

## Actual Behaviour

See above.

## Steps to Reproduce

1. Import a file like [test-import.csv](https://github.com/openfoodfoundation/openfoodnetwork/files/9984806/test-import.csv)
2. Check the product list afterwards

## Animated Gif/Screenshot

See above.

## Severity
bug-s4: it's annoying, but you can use it

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

### LINKED ISSUE #10101: [Product import] 500 error when a product has "0" in the unit value field
## Description
When importing a product file with "0" in the "units" column, a message "No error detected" is displayed to the user, but after clicking the "Save" button the import is stopped and an ugly error message appears on screen. 

## Expected Behavior
The import should be stopped when clicking the "Upload" button, with a proper error message, as "1 error prohibited this record from being saved: Unit value must be greater than 0."

## Actual Behavior
The import is allowed and the following request triggers a 500 Internal Server Error, leading to an ugly message being displayed to the user.

POST | https://staging.coopcircuits.fr/admin/product_import/save_data

## Steps to Reproduce
1. On import page (admin/product_import), import a csv file with "0" value in the "units" column. 
Test file: 
[RAHDK.csv](https://github.com/openfoodfoundation/openfoodnetwork/files/10155879/RAHDK.csv)
2. Click "upload" button, then "import" button, and see the message "No errors detected!", click Save

### After clicking the "Upload" button
Screen Shot 2022-12-05 at 17 08 55

### After clicking the "Save" button
Screen Shot 2022-12-05 at 17 09 11

## Workaround
Don't put 0.

## Severity
bug-s3: a feature is broken but there is a workaround

### PR DESCRIPTION
**Pull request by vviekk**
#### What? Why?

- Closes #9978 
- Closes #10101

Added validation to catch the following during product import:
1. Non numeric  `UNITS` value
   1.  Greater than zero condition for #10101 
1. `PRICE` validation.
   1. Can't be blank.
   2. Only accepts numeric values.
1. `ON_HAND` and `ON_DEMAND` validation.
   1. Both are only valid if at lest one of the conditions are met:
     1. on_demand = 1
     3. on_hand >= 0
1.  Changing the error message for `CATEGORY` to `admin.product_import.model.category_not_found` from `error_not_found_in_database`

#### What should we test?
- Test the above validations by uploading [test-import.csv](https://github.com/openfoodfoundation/openfoodnetwork/files/10833087/test-import.csv)

- Check for The following errors to be caught:
  1. Units can't be non-numeric or zero
  2. Price can't be blank or non-numeric
  3. on_hand and on_demand both can only be valid if either of the conditions are met: on_hand >=0 OR on_demand==1.
     1. If neither are met, then both are marked as invalid.

#### Follow ups
1. Add `conditional_blank` error message for the relation between `ON_HAND` and `ON_DEMAND` (similar to `variant_type` and `unit_type`).
5. The UI currently shows some fields populated by default data even when its empty in the file (Eg: `ON HAND` is set to 0 when empty) and empty for some cases when there should have been an entry (eg: Non numeric value for `PRICE` is not shown). Even though these are caught in validation with appropriate error messages, it could be confusing for the user if error message could be contradictory to what is show to the user. Eg: `Price` as incorrect value for non numeric data but empty cell is shown on the UI)
    1. update the rendering logic to reflect the actual data from the file. (Make sure not to break any existing usecase that could be tied)

### CODE REVIEW

**Comment by vviekk** [comment]
The failed checks seem to transient and unrelated to the change.

**Comment by vviekk** [comment]
Apologies for the delay on delivery.

File: app/models/product_import/entry_validator.rb [hunk 1]
```diff
@@ -189,6 +196,43 @@ def unit_fields_validation(entry)
...
+
+    def price_validation(entry)
+      return if is_numeric(entry.price)
+
+      if empty_or_placeholder_value(entry.price)
+        mark_as_invalid(entry, attribute: 'price',
+                               error: I18n.t('admin.product_import.model.blank'))
+      else
+        mark_as_invalid(entry, attribute: 'price',
+                               error: I18n.t('admin.product_import.model.incorrect_value'))
```
Comment (abdellani) [inline app/models/product_import/entry_validator.rb:211]: you can use a trinary  for the `error` attribute

Comment (vviekk) [reply app/models/product_import/entry_validator.rb:211]: Good suggestion. Thank you.
Will make the suggested change in the next revision with other comments (if any).

File: app/models/product_import/entry_validator.rb [hunk 2]
```diff
@@ -189,6 +196,43 @@ def unit_fields_validation(entry)
...
+      on_demand_value = entry.on_demand&.to_i
+
+      return if (on_hand_present_numeric && on_hand_value >= 0) ||
+                (on_demand_present_numeric && on_demand_value == 1)
+
+      mark_as_invalid(entry, attribute: 'on_hand',
+                             error: I18n.t('admin.product_import.model.incorrect_value'))
+      mark_as_invalid(entry, attribute: 'on_demand',
+                             error: I18n.t('admin.product_import.model.incorrect_value'))
+    end
```
Comment (dacook) [inline app/models/product_import/entry_validator.rb:230]: I don't see any advantage to validating these two fields in the same method, so could it be split into two?

File: app/models/product_import/entry_validator.rb [hunk 3]
```diff
@@ -189,6 +196,43 @@ def unit_fields_validation(entry)
...
+
+    def empty_or_placeholder_value(value)
+      !value&.present? || value.nil? || value.to_s.strip == '' || value.to_s.strip == "-"
```
Comment (dacook) [inline app/models/product_import/entry_validator.rb:233]: I think the first three conditions are equivalent to `#blank?` (https://blog.appsignal.com/2018/09/11/differences-between-nil-empty-blank-and-present.html ). Would you be able to try that?

**Review by dacook — CHANGES_REQUESTED** [review]
Thanks so much for this contribution @vviekk!

Firstly, I read [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/9978#issuecomment-1332331195) that "empty" fields are acceptable for ON HAND and ON DEMAND, but I don't see any specs that cover this case. It would be great to code this in the specification to ensure we know for the future.
Would you be able to add specs for this? I think we should have more unit testing in `entry_validator_spec.rb`, but adding to the system spec `product_import_spec.rb` for now would be fine if that's easier.

I've also left some other comments, which are not essential but might be worth addressing if possible.

File: app/models/product_import/entry_validator.rb (hunk 2 above)
Comment (vviekk) [reply app/models/product_import/entry_validator.rb:230]: I initially tried having separate validation methods for them. 
But as we had to read both of the values in both the methods(for the relation between on_hand and on_validation), it seemed repetitive.

Eg: 

```
def on_hand_validation(entry)
  on_hand_present_numeric = !empty_or_placeholder_value(entry.on_hand) &&
                            is_numeric(entry.on_hand)
  on_hand_value = entry.on_hand&.to_i

  return if on_hand_present_numeric && on_hand_value >= 0 ||
            (!empty_or_placeholder_value(entry.on_demand) &&
             entry.on_demand&.to_i == 1)

  mark_as_invalid(entry, attribute: 'on_hand',
                         error: I18n.t('admin.product_import.model.incorrect_value'))
end

def on_demand_validation(entry)
  on_demand_present_numeric = !empty_or_placeholder_value(entry.on_demand) &&
                              is_numeric(entry.on_demand)
  on_demand_value = entry.on_demand&.to_i

  return if on_demand_present_numeric && on_demand_value == 1 ||
            (!empty_or_placeholder_value(entry.on_hand) &&
             entry.on_hand&.to_i >= 0)

  mark_as_invalid(entry, attribute: 'on_demand',
                         error: I18n.t('admin.product_import.model.incorrect_value'))
end

```

Having said that, i've reduced the validation to the relation between the fields,i.e, if at least one of them has a valid value, other one is ignored regardless of its value.

Eg: **Ignoring cases like entry 2 and 3**
| Entry     | on_hand | on_demand | Output | Ideal output |
|-----------|---------|-----------|-----------------|--|
| Entry 1   | 10      | 1         | Valid           | Valid|
| Entry 2   | -5      | 1         | Valid           | Invalid on_hand|
| Entry 3   | foo     | 1         | Valid           | Invalid on_hand|
| Entry 4   | 10      | 0         | Valid           | Valid|
| Entry 5   | -5      | 0         | Invalid           |Invalid|
| Entry 6   | foo     | 0         | Invalid           |Invalid|
| Entry 7   |          | 1         | Valid           |Valid|
| Entry 8   | -10     |           | Invalid           |Invalid|
| Entry 9   |          |           | Invalid         |Invalid|
| Entry 10  | 5     | 1         | Valid         |Valid|
| Entry 11  | 5     |           | Valid         |Valid|
| Entry 12  | -5      |           | Invalid         |Invalid|

File: app/models/product_import/entry_validator.rb (hunk 3 above)
Comment (vviekk) [reply app/models/product_import/entry_validator.rb:233]: You are correct.
Thank you. Will make the suggested changes.

File: app/models/product_import/entry_validator.rb (hunk 2 above)
Comment (vviekk) [reply app/models/product_import/entry_validator.rb:230]: Also while adding tests, I learned that they have to be mutually exclusively enabled for importing type as "Inventories".
Added more test covering this case as well.

**Comment by vviekk** [comment]
@dacook 
Thank you for the comments!
Regarding the specs for empty values for on_hand and on_demand, most were already a part of the tests but I have also modified one of the existing specs to include more.
I've also added more test cases for non numeric and negative values.

File: app/models/product_import/entry_validator.rb (hunk 1 above)
Comment (vviekk) [reply app/models/product_import/entry_validator.rb]: Could not use tertiary operator here as the code style needed it be in one line and when I did that, it triggered another linter error for line too long. But made changes to take away code repetition as suggested.

File: app/models/product_import/entry_validator.rb (hunk 2 above)
Comment (dacook) [reply app/models/product_import/entry_validator.rb:229]: Ah, thanks for responding, I didn't pick up on the relationship between the two.

**Review by dacook — APPROVED** [review]
Thanks for updating and increasing the specs. 
We now have a lot of tests covering every case which is good, although it may be harder to maintain than simpler unit tests. All good from my end.

I checked the failing spec which seems unrelated and must be intermittent. I've moved to the Code Review column and will await one further review before we proceed to manual testing and merge.

**Review by jibees — COMMENTED** [review]
This looks good to me and here's my 2 cents:

 - I think that finally those two commits should be squashed (the second one address review comments),
 - I'm also concerned about instance that do use `,` (comma) as a decimal separator (like we do in France): can the document be imported if it use `,` as decimal separator for price for example? 

Thanks a lot!!

**Comment by vviekk** [comment]
@jibees 
Great point!
Just so I understand correctly, price value in the import file could look something like: `2,50`, correct?
And we are to accept that as a valid value.

**Comment by jibees** [comment]
> Just so I understand correctly, price value in the import file could look something like: `2,50`, correct?

Yes. 

> And we are to accept that as a valid value.

I guess so. 


@RachL do you confirm?

**Comment by vviekk** [comment]
@jibees 
Another quick question: 
If we are doing it for France, should we also account for other number systems? Like done in [price_parser](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/assets/javascripts/admin/utils/services/price_parser.js.coffee)?

| Country/Region | Example Number | English/Computer Decimal Representation |
| --- | --- | --- |
| United States, United Kingdom, Canada, and many other English-speaking countries | 1,234.56 | 1234.56 |
| France, Belgium, and some other European countries | 1 234,56 | 1234.56 |
| Switzerland and Liechtenstein | 1'234'567.89 | 1234567.89 |
| Italy | 1.234.567,89 | 1234567.89 |
| India | 1,23,456.78 | 123456.78 |

**Comment by vviekk** [comment]
#### Current status
Sample import - [test-import.csv](https://github.com/openfoodfoundation/openfoodnetwork/files/10872158/test-import.csv)

If we import a file with price value set as `4,20` (see last entry in the file above), the import converts it into `420`, instead of `4.20`.
And the second last with value `4,204.23` is converted to `4204.23`.

image
Seems like we currently only support the English decimal system. (Which simply removes any comma delimiter present)

@jibees @RachL 
I think we assume users stick to the point decimal system while importing.
If they are not, there might be a bigger issue.

**Comment by jibees** [comment]
Ok seems like you're absolutely right: we don't manage `,` decimal separator for CSV importing, see #9474. So, unless @RachL disagree, I think this PR does improve UX and does not degrade it, so, I'd approve it.

**Comment by RachL** [comment]
thanks @vviekk yes let's move forward :+1:

**Review by jibees — APPROVED** [review]
👍 
Thanks for this discussion, let's move forward!

**Comment by drummer83** [comment]
Hi @vviekk,
I finally had a chance to test this. Thank you very much.

First of all I have to say that the behaviour seems to have changed since #9978 had been created. Currently (before your PR) some cases are failing without any error message during import. The import just hangs here at 0 %. Maybe the situation is different on your local setup?

image

However I compared the situation before and after your PR for units and price.

## Units

| Units value | Before the PR | After the PR |
| - | - | - |
| empty | import hangs without error message :x: | import hangs without error message :x: |
| 0 | import hangs without error message :x: | import hangs without error message :x: |
| -3 | import hangs without error message :x: | import hangs without error message :x: |
| ?/*% | import hangs without error message :x: | import hangs without error message :x: |
| 0.12 | ok ✔️ | Error: Units incorrect value :x: ❗  |
| -0.12 | import hangs without error message :x: | import hangs without error message :x: |
| "0,34" | import hangs without error message :x: | import hangs without error message :x: |
| 5 | ok ✔️ | ok ✔️ |

## Price

| Price value | Before the PR | After the PR |
| - | - | - |
| empty | import hangs without error message :x: | import hangs without error message :x: |
| 0 | ok ✔️ | ok ✔️ |
| -3 | import hangs without error message :x: | import hangs without error message :x: |
| ?/*% | no error, price = 0 :x: | Error: Price incorrect value (displayed as empty) ✔️ |
| 0.12 | ok ✔️ | ok ✔️ |
| -0.12 | import hangs without error message :x: | import hangs without error message :x: |
| "0,34" | ok ✔️ | Error: Price incorrect value (displayed as 034) ❓|
| 5 | ok ✔️ | ok ✔️ |

# Results
I did not yet test the 'on hand' and 'on demand' behaviour. However I think we should fix the regression of not accepting 0.12 as a correct value for units. 1.12 works by the way.

Could you have another look, please?

**Comment by vviekk** [comment]
@drummer83 thank you.
Will do!

**Comment by vviekk** [comment]
@drummer83 Thanks for the tests.

>The import just hangs here at 0 %. Maybe the situation is different on your local setup?

I suspect its the headers in the csv, it perhaps expects some specific columns and/or order. Please try with the following test files.
1. Please create an enterprise with the name `test` before starting the import test with the following files.

# I see the following on my local:

### Unit:  [Import test - unit.csv](https://github.com/openfoodfoundation/openfoodnetwork/files/10988505/Import.test.-.unit.csv)

#### For current prod:
image
1. All the values are accepted except empty.

#### For prod + this PR:
image
1. Entries with empty, zero, negative values, non-numeric are invalidated.
2. For entry number 6: I had put in a check to invalidate 0 or negative unit values (in favor of #10101). That check converts the value into int (0.12 -> 0). To fix this I'll convert it into float for the check. -- Fix in this PR
3. For entry number 9:  The system is trying to see if the string "0,34" is a valid number (float value). As our system only expect decimal number system, this value is invalidated.
  1. Even if we don't invalidate it here, the value is likely invalidated while saving the data OR Simply converted to `34` instead of the intended value of `0.34` (decimal representation).
  2. I suspect the later is the case.
### Price: [import test - price.csv](https://github.com/openfoodfoundation/openfoodnetwork/files/10988553/impoer.test.-.price.csv)

#### For current prod:
image
1. All the values are accepted except empty.

#### For prod + this PR
image
1. All Numeric values are accepted except empty and non-numeric value(entry 5 - It shows empty in the table but error message is correct).
image
1. Entry number 9 has value 0.34 in the import file unlike the import file for unit value test above.
 
**Do we want to add a check for negative value?**



----------------------


## Summary:
1. The import hangs at times as it might expect the file to be in certain format.
2. For unit - All the negative, 0 and non-numeric values are invalidated.
    1. `0.12` is invalidated because the check converts it into `0` (int) and fails the condition of `>0`. Need to make change such that the check is  `0.12>0`.
3. Price accepts all numeric values even negative and rejects all non-numeric values.
4. The system does not support Comma decimal system. [Discussion here](https://github.com/openfoodfoundation/openfoodnetwork/pull/10461#issuecomment-1451939400)


Let me know if there's any confusion or queries. :)

**Comment by filipefurtad0** [comment]
Hey @vviekk ,

Before this PR several cases were displaying this validation:

image


After this PR, I've checked that
- Non numeric UNITS value triggers the correct error
- numeric UNITS are accepted,  if greater than 0

image

- PRICE Can't be blank.
- PRICE Only accepts numeric values.

image

- ON_HAND and ON_DEMAND are only valid if at lest one of the conditions are met :green_circle: 
    on_demand = 1
    on_hand >= 0

Changing the error message for CATEGORY to admin.product_import.model.category_not_found from error_not_found_in_database

image


Awesome! Good to go :muscle:
